cloudwatch/cloudwatch.go

```diff
@@ -179,6 +193,28 @@ func NewOutputPlugin(config OutputPluginConfig) (*OutputPlugin, error) {
 		return nil, err
 	}
 
+	var data TaskMetadata
+	// check if it is running in ECS
+	ecsTaskMetadataEndpointV3 := os.Getenv("ECS_CONTAINER_METADATA_URI")
+	if ecsTaskMetadataEndpointV3 != "" {
+		metadata, err := getECSMetadata(ecsTaskMetadataEndpointV3)
+		if err != nil {
```

Comment:
This is not what I meant with the code re-organization; I don't think its what David meant either.

I want all of this functionality to be inside of the `getECSMetadata` function. That function should set the metadata on the OutputPlugin or return an error. That way all of the logic is organized into one place.

I want the `getECSMetadata` function to be called in the `AddRecord` function like it used to be. That way if it fails, it can be retried. If you do it here, then if it fails Fluent Bit will not start.

Reply:
This is actually close to what I meant, but you make a good point about the retry. That should be reconsidered with your points in mind. I think `getECSMetadata` can be a method on the `OutputPlugin` struct (like it was?). Then it can set the `TaskMetadata` variables directly and basically return err or nil to trigger a retry.

Reply:
Got it. Then I can make the method on the `OutputPlugin `struct  and call it in `AddRecord`. I will re-implement it